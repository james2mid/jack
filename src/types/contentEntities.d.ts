export interface ContentEntities {
  /** The text of the content with placeholders. */
  text: string
  /** URLs within within the content. (:) */
  urls: string[]
  /** Hashtags used within the content. (#) */
  hashtags: string[]
  /** Cashtags used within the content. ($) */
  cashtags: string[]
  /** The usernames mentioned in the content. (@) */
  usernames: string[]
  /** The id of each user mentioned within the content. */
  userIds: string[]
}