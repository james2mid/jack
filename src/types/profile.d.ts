import { ContentEntities } from "./contentEntities";

type Without<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>
type BioEntities = Without<ContentEntities, 'userIds'>

/** The full profile scraped from `twitter.com/username`. */
export interface FullProfile extends Profile {
  /** The date which the user created their twitter account. */
  joinedAt: Date
  /** The location text chosen by the user to be shown on their profile. */
  geo?: string,
  /** The URL of the website chosen by the user to be visible on their profile. */
  websiteUrl?: string,
  /** The colour (as hex with '#') chosen by the user to be used on their profile page. */
  color: string
}

/** The partial profile scraped from `twitter.com/i/profiles/popup`. */
export interface Profile {
  /** Twitter's unique ID for this user. */
  userId: string
  /** The URL which links to the user's banner image. */
  bannerUrl?: string
  /** The URL which links to the user's avatar image. */
  avatarUrl?: string
  /** The full name of the user. */
  name: string
  /** The Twitter handle or screen name. */
  username: string,
  /** Whether the user has opted to make their tweets visible to only those which have a mutual following. */
  isProtected: boolean,
  /** Holds the text and entities within a user's bio. */
  bio: BioEntities,
  /** The time when this user's properties were last updated. */
  lastUpdated: Date,
  /** Properties of a user which change often. */
  stats: {
    /** The number of tweets the user has posted. */
    tweetCount: number,
    /** The number of users this account is following. */
    followingCount: number,
    /** The number of accounts following this user. */
    followersCount: number
  },
  /** The HTML which was parsed to generate this profile object. */
  html: string
}