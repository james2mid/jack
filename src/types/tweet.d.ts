import { ContentEntities } from "./contentEntities";

export interface Tweet {
  /** Twitter's unique ID for this tweet. */
  tweetId: string
  /** The unique Twitter ID for the user posting this tweet. */
  userId: string
  /** The username of the user who posted this tweet. */
  username: string
  /** The time the tweet was posted. */
  timestamp: Date
  /** Holds tweet text and entities. */
  content: ContentEntities
  /** The id of the tweet which is the root of this conversation. */
  conversationId?: string
  /** The id of the tweet which this one is replying to. */
  mentionedTweet?: string
  /** The HTML of the scraped tweet. */
  html: string
  /** The time when the tweet's `stats` properties were last updated. */
  lastUpdated: Date
  /** An object containing values about the tweet which change with time. */
  stats: {
    /** The number of replies to this tweet. */
    replyCount: number
    /** The number of times this tweet has been retweeted. */
    retweetCount: number
    /** The number of times this tweet has been liked. */
    likeCount: number
  }
}