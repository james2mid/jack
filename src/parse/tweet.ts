import cheerio from 'cheerio'
import { parseTweetEntities } from './entities';
import { toCheerio, Cheerioable } from './common';
import { minify } from 'html-minifier'
import { Tweet } from '../types/tweet';

/** Parses the first tweet found using the selector on the provided cheerioable. */
export function parseTweet (tweet: Cheerioable, selector: string = '.tweet'): Tweet {
  
  const $tweet = toCheerio(tweet, selector).first()

  if ($tweet.length === 0) {
    throw new Error('No tweet found in `tweet` paremeter')
  }

  const tweetId: string  = $tweet.attr('data-tweet-id')
  const userId: string = $tweet.attr('data-user-id')
  const username: string = $tweet.attr('data-screen-name')
  
  const timestamp: Date = (() => {
    const msEpoch = $tweet.find('span._timestamp').attr('data-time-ms')
    return new Date(parseInt(msEpoch))
  })()

  const mentionedTweet: string | undefined = (() => {
    const $quotedTweet = $tweet.find('.QuoteTweet-link')
    if ($quotedTweet.length > 0) {
      // get tweet id of quoted tweet
      return $quotedTweet.attr('data-conversation-id')
    } else {
      return undefined
    }
  })()

  const content = parseTweetEntities($tweet, '.tweet-text')

  const html: string = minify(cheerio.load($tweet[0]).html($tweet))

  const lastUpdated = new Date()

  const stats = {
    replyCount: parseInt($tweet.find('.ProfileTweet-action--reply [data-tweet-stat-count]').attr('data-tweet-stat-count')),
    retweetCount: parseInt($tweet.find('.ProfileTweet-action--retweet [data-tweet-stat-count]').attr('data-tweet-stat-count')),
    likeCount: parseInt($tweet.find('.ProfileTweet-action--favorite [data-tweet-stat-count]').attr('data-tweet-stat-count')),
  }

  return { tweetId, userId, username, timestamp, mentionedTweet, content, html, lastUpdated, stats }
}
