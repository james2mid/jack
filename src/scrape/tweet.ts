import axios from "axios";
import { Tweet } from "../types/tweet";
import { validId } from "./common";
import { parseTweet } from "../parse/tweet";

/** Scrapes a single tweet by its id. */
export async function getTweet (tweetId: string): Promise<Tweet> {
  if (!validId(tweetId)) {
    throw new Error(`This tweet id is not a valid one: '${tweetId}'`)
  }

  const url = `https://twitter.com/i/web/status/${tweetId}`

  const response = await axios.get(url)
  const html = response.data
  return parseTweet(html, '.permalink-tweet')
}