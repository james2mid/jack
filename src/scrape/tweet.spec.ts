import { getTweet } from "./tweet";
import { Tweet } from "../types/tweet";

describe('getTweet', () => {

  const tweets = [
    '378675940261261312',
    '1064437357359726592',
    '1064472353453867008',
    '1067373627266555904'
  ]

  const deletedTweets = [
    '1064484063510179840'
  ]

  it('rejects immediately for an invalid tweetId', () => {
    const promise = getTweet('invalid id')
    return expect(promise).rejects.toThrowError()
  }, 5)

  it('rejects with 404 for deleted tweets', async () => {
    const promises = deletedTweets.map(x => getTweet(x))
    await Promise.all(promises.map(x => expect(x).rejects.toMatchObject({
      response: {
        status: 404
      }
    })))
  })


  let results: Tweet[]

  beforeAll(async () => {
    results = await Promise.all(tweets.map(x => getTweet(x)))
  }, 1000 * tweets.length)

  it('gets the expected objects', () => {
    results.forEach(tweet => {
      expect(tweet).toMatchSnapshot({
        html: expect.any(String),
        lastUpdated: expect.any(Date),
        stats: {
          likeCount: expect.any(Number),
          replyCount: expect.any(Number),
          retweetCount: expect.any(Number)
        }
      })
    })
  })

})