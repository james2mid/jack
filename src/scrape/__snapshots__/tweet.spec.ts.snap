// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTweet gets the expected objects 1`] = `
Object {
  "content": Object {
    "cashtags": Array [],
    "hashtags": Array [],
    "text": "I wish the twitter team the best of luck and trust that they will be successful in continuing to develop this important communication tool.",
    "urls": Array [],
    "userIds": Array [],
    "usernames": Array [],
  },
  "html": Any<String>,
  "lastUpdated": Any<Date>,
  "mentionedTweet": undefined,
  "stats": Object {
    "likeCount": Any<Number>,
    "replyCount": Any<Number>,
    "retweetCount": Any<Number>,
  },
  "timestamp": 2013-09-14T00:25:54.000Z,
  "tweetId": "378675940261261312",
  "userId": "14",
  "username": "noah",
}
`;

exports[`getTweet gets the expected objects 2`] = `
Object {
  "content": Object {
    "cashtags": Array [],
    "hashtags": Array [],
    "text": "“Just buy more if it goes lower”",
    "urls": Array [],
    "userIds": Array [],
    "usernames": Array [],
  },
  "html": Any<String>,
  "lastUpdated": Any<Date>,
  "mentionedTweet": undefined,
  "stats": Object {
    "likeCount": Any<Number>,
    "replyCount": Any<Number>,
    "retweetCount": Any<Number>,
  },
  "timestamp": 2018-11-19T08:36:56.000Z,
  "tweetId": "1064437357359726592",
  "userId": "1032188509",
  "username": "Crypto_Bitlord",
}
`;

exports[`getTweet gets the expected objects 3`] = `
Object {
  "content": Object {
    "cashtags": Array [],
    "hashtags": Array [
      "smm",
      "socialmedia",
      "marketing",
    ],
    "text": "Who Were LinkedIn’s 2018 Top Voices for Marketing and Social Media? \${:0}… \${#0} \${#1} \${#2}",
    "urls": Array [
      "http://buff.ly/2A2j3BH",
    ],
    "userIds": Array [],
    "usernames": Array [],
  },
  "html": Any<String>,
  "lastUpdated": Any<Date>,
  "mentionedTweet": undefined,
  "stats": Object {
    "likeCount": Any<Number>,
    "replyCount": Any<Number>,
    "retweetCount": Any<Number>,
  },
  "timestamp": 2018-11-19T10:56:00.000Z,
  "tweetId": "1064472353453867008",
  "userId": "971620273",
  "username": "SocialMediaKE",
}
`;

exports[`getTweet gets the expected objects 4`] = `
Object {
  "content": Object {
    "cashtags": Array [],
    "hashtags": Array [],
    "text": "Follow \${@0} for some pretty interesting threads",
    "urls": Array [],
    "userIds": Array [
      "323993141",
    ],
    "usernames": Array [
      "MisterCh0c",
    ],
  },
  "html": Any<String>,
  "lastUpdated": Any<Date>,
  "mentionedTweet": "1057614142905991168",
  "stats": Object {
    "likeCount": Any<Number>,
    "replyCount": Any<Number>,
    "retweetCount": Any<Number>,
  },
  "timestamp": 2018-11-27T11:04:38.000Z,
  "tweetId": "1067373627266555904",
  "userId": "913367713837408257",
  "username": "cvrbonbased",
}
`;
